### ISSUE TITLE
New weight still produce wrong result

### ISSUE THREAD

**Issue by AtlantixJJ** [mseitzer/pytorch-fid#17]
Using the updated weights still get wrong result. See this repo https://github.com/AtlantixJJ/PytorchInceptionV3 for detail.

Run this code (need to store some image in data/cifar10_test, or go to the repo above):

```python
"""
A script to test Pytorch and Tensorflow InceptionV3 have consistent behavior.
"""
import sys, argparse, os, pathlib
sys.path.insert(0, ".")
import numpy as np
import tensorflow as tf
import torch
from inception_modified import inception_v3
from PIL import Image

parser = argparse.ArgumentParser()
parser.add_argument("--load_path", default="", help="The path to changed pytorch inceptionv3 weight. Run change_statedict.py to obtain.")
args = parser.parse_args()

def check_or_download_inception(inception_path):
    ''' Checks if the path to the inception file is valid, or downloads
        the file if it is not present. '''
    INCEPTION_URL = 'http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz'
    if inception_path is None:
        inception_path = '/tmp'
    inception_path = pathlib.Path(inception_path)
    model_file = inception_path / 'classify_image_graph_def.pb'
    if not model_file.exists():
        print("Downloading Inception model")
        from urllib import request
        import tarfile
        fn, _ = request.urlretrieve(INCEPTION_URL)
        with tarfile.open(fn, mode='r') as f:
            f.extract('classify_image_graph_def.pb', str(model_file.parent))
    return str(model_file)

def torch2numpy(x):
    return x.detach().cpu().numpy().transpose(0, 2, 3, 1)

torch.backends.cudnn.benchmark = True
torch.manual_seed(1)
torch.cuda.manual_seed(1)

data_dir = "data/cifar10_test/"
imgs_pil = [Image.open(open(data_dir + s, "rb")).resize((299,299), Image.BILINEAR) for s in os.listdir(data_dir)]
imgs = [np.asarray(img).astype("float32") for img in imgs_pil]
x_arr = np.array(imgs)
x_arr_tf = x_arr
# TF InceptionV3 graph use [0, 255] scale image
feed = {'FID_Inception_Net/ExpandDims:0': x_arr_tf}
# This is identical to TF image transformation
x_arr_torch = x_arr / 255. #(x_arr - 128) * 0.0078125
x_torch = torch.from_numpy(x_arr_torch.transpose(0, 3, 1, 2)).float().cuda()

model = inception_v3(pretrained=True, aux_logits=False, transform_input=False)
if len(args.load_path) > 1:
    # default: pretrained/inception_v3_google.pth
    print("=> Get changed weight from %s" % args.load_path)
    try:
        model.load_state_dict(torch.load(args.load_path))
    except RuntimeError:
        pass
model.cuda()
model.eval()

if x_torch.size(2) != 299:
    import torch.nn.functional as F
    x_torch = F.interpolate(x_torch,
            size=(299, 299),
            mode='bilinear',
            align_corners=False)
features = model.get_feature(x_torch)
feature_pytorch = features[-1].detach().cpu().numpy()
if len(feature_pytorch.shape) == 4:
    feature_pytorch = feature_pytorch[:, :, 0, 0]

inception_path = check_or_download_inception("pretrained")
with tf.gfile.FastGFile("pretrained/classify_image_graph_def.pb", 'rb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    _ = tf.import_graph_def( graph_def, name='FID_Inception_Net')
    
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)

layername = "FID_Inception_Net/pool_3:0"
layer = sess.graph.get_tensor_by_name(layername)
ops = layer.graph.get_operations()
for op_idx, op in enumerate(ops):
    for o in op.outputs:
        shape = o.get_shape()
        if shape._dims != []:
            shape = [s.value for s in shape]
            new_shape = []
            for j, s in enumerate(shape):
                if s == 1 and j == 0:
                    new_shape.append(None)
                else:
                    new_shape.append(s)
            # print(o.name, shape, new_shape)
            o.__dict__['_shape_val'] = tf.TensorShape(new_shape)

tensor_list = [n.name for n in tf.get_default_graph().as_graph_def().node]

target_layer_names = ["FID_Inception_Net/Mul:0", "FID_Inception_Net/conv:0", "FID_Inception_Net/pool_3:0"]
target_layers = [sess.graph.get_tensor_by_name(l) for l in target_layer_names]

sess.run(tf.global_variables_initializer())
res = sess.run(target_layers, feed)
x_tf = res[0]
feature_tensorflow = res[-1][:, 0, 0, :]

print("=> Pytorch pool3:")
print(feature_pytorch[0][:6])
print("=> Tensorflow pool3:")
print(feature_tensorflow[0][:6])
print("=> Mean abs difference")
print(np.abs(feature_pytorch - feature_tensorflow).mean())

def get_tf_layer(name):
    return sess.run(sess.graph.get_tensor_by_name(name + ':0'), feed)
```

result:

```
=> Pytorch pool3:
[0.42730308 0.00819586 0.27243498 0.2880235  0.10205843 0.05626289]
=> Tensorflow pool3:
[0.13085267 0.5260418  0.22931635 0.02919772 0.2439549  0.50965333]
=> Mean abs difference
0.34774715
```

**Comment by mseitzer** [comment]
Hi, 

I made a similar script to yours, and I get nearly identical feature maps between Pytorch and Tensorflow for identical inputs.

I don't think that you use the Inception model in this repository with the latest changed by me, but the original Inception from torchvision. This Inception implementation does not correspond to the one used in Tensorflow's FID. Could you please run your tests again, but with the newest model (and weights) provided here?

**Comment by AtlantixJJ** [comment]
I downloaded your model and run the script with command `python test_inception.py --load_path <your_model>`
So I used your weight.
Could you please provide your script?

**Comment by AtlantixJJ** [comment]
What's more, in my result the difference is concentrated on edge. Perhaps the padding is not consistent.
[tf_pytorch_error.pdf](https://github.com/mseitzer/pytorch-fid/files/3250686/tf_pytorch_error.pdf)

**Comment by AtlantixJJ** [comment]
Sorry you are right. I didn't use your model. Now my results are 

```
=> Pytorch pool3:
[0.13110925 0.5254472  0.22908828 0.02930989 0.24280292 0.51165456]
=> Tensorflow pool3:
[0.1308305  0.52602327 0.22930455 0.02920746 0.24395223 0.5096492 ]
=> Mean abs difference
0.0011506503
```

I believe your repo is correct. But I am wondering why the official Pytorch Inception is different. Is the official network architecture different from yours? Because I ported the Tensorflow weights directly and the weights I used should be exactly same.

**Comment by mseitzer** [comment]
Yes, I discovered that there are some minor differences between the Pytorch implementation and the Inception model used by FID. There are different Inception graphs/weights published. If you open the graph in Tensorboard, you can see some differences.

You can see the necessary changes to the model in this commit: https://github.com/mseitzer/pytorch-fid/commit/f64228cbad0988a6e134013c396e80a09692df01. The changes are marked in comments beginning with `Patch`.

By the way, I think you can get even lower mean difference if you use exactly the same input tensor for both Tensorflow and Pytorch. So resize your image only once, with PIL, and then skip the resize of the Tensorflow graph (with input key `FID_Inception_Net/Mul:0`, but be aware that this requires your inputs to in range `[-1, 1]`).

**Comment by AtlantixJJ** [comment]
Yes. It is now on 1e-5. Thank you for your work!

**Comment by AtlantixJJ** [comment]
@mseitzer But do you really think we should use the result exactly the same as TTUR? According to your patch, they do not use the correct Inception code. Maybe the official weight of Pytorch is better for evaluation.

**Comment by mseitzer** [comment]
Well, I don't think there is much choice here. The authors of FID chose this version of Inception, and to stay comparable, everyone has to use this version now. I guess the FID authors just took the same model Inception score uses: https://github.com/openai/improved-gan/blob/master/inception_score/model.py

Also, this Inception model is not "wrong", it just does not directly correspond to the paper. Its feature maps should still contain what the authors of FID and Inception scores want to measure.

But I *do* think it is a bit concerning that FID score gives wildly different results if you use a different version of Inception. In my opinion, the metric should not be sensitive to that.

**Comment by mseitzer** [comment]
Fixed by https://github.com/mseitzer/pytorch-fid/pull/16

**Comment by vibss2397** [comment]
hey @AtlantixJJ can you tell me what change you made to the test_inception.py script which led you to get better results?
